File: lib/easy/middle_linked_list_node.dart
import 'package:dark_coding_challenges/easy/list_node.dart';

ListNode<T>? middleLinkedListNode<T>(ListNode<T>? head) {
  ListNode<T>? slow = head;
  ListNode<T>? fast = head;
  while (fast != null && fast.next != null) {
    slow = slow?.next;
    fast = fast.next?.next;
  }
  return slow;
}

File: easy/middle_linked_list_node.go
package easy

func MiddleLinkedListNode[T ComparableOrdered](head *ImprovedListNode[T]) *ImprovedListNode[T] {
	if head == nil {
		return nil
	}
	slow := head
	fast := head
	for fast != nil && fast.Next != nil {
		slow = slow.Next
		fast = fast.Next.Next
	}
	return slow
}
